Buildstep: add type-checking on description et al

jaredgrubb opened the pull request:
I was porting some ShellCommand derived classes to BuildStep + ShellMixin and hit some type problems around "description", "descriptionDone", and "descriptionSuffix". This patch moves all the ensure-list code into the BuildStep base (where the attributes are set).

I'll port this forward to nine as well.

sa2ajj commented:
```
[ERROR]
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/Twisted-15.3.0-py2.7-linux-x86_64.egg/twisted/internet/defer.py", line 1128, in _inlineCallbacks
    result = g.send(result)
  File "/home/travis/build/buildbot/buildbot/master/buildbot/process/buildstep.py", line 291, in startStep
    self._step_status.setText(self.describe(False))
  File "/home/travis/build/buildbot/buildbot/master/buildbot/steps/http.py", line 171, in describe
    return self.description.split()
exceptions.AttributeError: 'list' object has no attribute 'split'

buildbot.test.unit.test_steps_http.TestHTTPStep.test_404
buildbot.test.unit.test_steps_http.TestHTTPStep.test_POST
buildbot.test.unit.test_steps_http.TestHTTPStep.test_basic
buildbot.test.unit.test_steps_http.TestHTTPStep.test_header
buildbot.test.unit.test_steps_http.TestHTTPStep.test_params_renderable
-------------------------------------------------------------------------------
Ran 3106 tests in 29.656s

FAILED (skips=8, errors=5, successes=3093)
```

sa2ajj commented on master/buildbot/steps/source/base.py:
```diff
@@ -94,7 +94,25 @@ def __init__(self, workdir=None, mode='update', alwaysUseLatest=False,
+            "export": "export"}
+
+        if not description:
+            description = [
```
why split the line??

jaredgrubb replied:
That's a copy-paste thing .. I think someone was trying to keep the line from being too long?

sa2ajj replied:
Do you think you could fix it?  It does not help with too long lines, but it makes things [a bit] less readable.

sa2ajj commented on master/buildbot/steps/source/base.py:
```diff
@@ -94,7 +94,25 @@ def __init__(self, workdir=None, mode='update', alwaysUseLatest=False,
+            descriptionDone = [
+                descriptionDones_for_mode.get(mode, "update")]
+        if not descriptionSuffix:
+            descriptionSuffix = codebase or None  # want None in lieu of ''
```
why not a list?

jaredgrubb replied:
Yes, I think this should be a list. Thanks.

sa2ajj replied:
You did not update this to list, are going to?

sa2ajj commented on master/buildbot/steps/source/base.py:
```diff
@@ -94,7 +94,25 @@ def __init__(self, workdir=None, mode='update', alwaysUseLatest=False,
+            description = [
+                descriptions_for_mode.get(mode, "updating")]
+        if not descriptionDone:
+            descriptionDone = [
```
ditto

jaredgrubb commented:
Alright I've addressed the issues. The HTTP step steps now pass (turns out I had to install 'txrequests' in order to run them; they were being skipped on my machine).

jaredgrubb commented:
Yes, you're right ... I had fixed those issues on my nine branch, but not on THIS one. So, I think we're good now. (I'll post nine one next)
